### PR TITLE
Enumerable

### PR DESCRIPTION
**Pull request by apurvis**
OK @jarthod I figured it out.  Supersedes the other PR but given that this might be considered a breaking change let me know what you want to do about versioning (i will leave both PRs open)

### CODE REVIEW

File: lib/survey_gizmo/resource.rb [hunk 1]
```diff
@@ -33,47 +34,41 @@ module ClassMethods
       def all(conditions = {}, _deprecated_filters = {})
         conditions = merge_params(conditions, _deprecated_filters)
         fail ':all_pages and :page are mutually exclusive' if conditions[:page] && conditions[:all_pages]
+        fail 'Block only makes sense with :all_pages' if block_given? && !conditions[:all_pages]
```
Comment (jarthod) [inline lib/survey_gizmo/resource.rb]: You don't need this any more

File: lib/survey_gizmo/resource.rb [hunk 2]
```diff
@@ -33,47 +34,41 @@ module ClassMethods
...
-          next unless v && instance_methods.include?(k)
-          collection.each { |c| c[k] ||= v }
-        end
-
-        # Sub questions are not pulled by default so we have to retrieve them manually
-        # SurveyGizmo claims they will fix this bug and eventually all questions will be
-        # returned in one request.
-        if self == SurveyGizmo::API::Question
-          collection += collection.map { |question| question.sub_questions }.flatten
+        collection = []
```
Comment (jarthod) [inline lib/survey_gizmo/resource.rb]: You don't need this neither

File: README.md [hunk 3]
```diff
@@ -45,12 +51,24 @@ SurveyGizmo.configure do |config|
...
-surveys = SurveyGizmo::API::Survey.all
+# Iterate over your surveys directly with the iterator
+SurveyGizmo::API::Survey.all(all_pages: true).each do |survey|
+  do_something_with(survey)
+end
 # Retrieve ALL your surveys (handle pagination for you)
-surveys = SurveyGizmo::API::Survey.all(all_pages: true)
+surveys = SurveyGizmo::API::Survey.all(all_pages: true).to_a
+# Retrieve the 1st page of your surveys
+surveys = SurveyGizmo::API::Survey.all(page: 1).to_a
```
Comment (jarthod) [inline README.md]: I don't think you should add `.to_a` to `.all` examples, it is best practice to keep the iterators and iterate it after. This way we only query what we need. (if surveys is never used for example, no call will be made)

**Comment by jarthod** [comment]
Yeah this change might break for people doing uncommon things with `.all` but it should be ok for most (calling .each) So I think this can go to 4.2 with a small note in the readme.

**Comment by jarthod** [comment]
Good job BTW!

**Comment by apurvis** [comment]
it actually was surprisingly easy...  however i will say i'm not even sure if this change wouldn't have broken my code so maybe it's a 5.0?

i was trying to think of a way around that but i can't really.  i'm also just not afraid of bumping version numbers lol.

**Comment by apurvis** [comment]
i feel like this is a common idiom (it was my idiom)

```ruby
page = 1
responses = Response.all(page: page)
while !responses.empty?
  do_some_stuff
  page += 1
  responses = Response.all(page: page)
end
```

and this change would break that.

**Comment by apurvis** [comment]
ok i addressed yr comments - definitely better to not use `.to_a` in the examples though I did leave one example to indicate what a `.to_a` call would do.

**Comment by apurvis** [comment]
i put some proposed 5.x release notes in the README... not sure though if we should bother fixing the deprecated two hash versions of `.all` and `.first` because those will really break people's code.

on the one hand, bumping the major version gives license to break things.  on the other, it still kind of sucks to break people's code.

**Comment by apurvis** [comment]
We could also merge the other PR into a 4.2 release and then this into a 5.0

**Comment by apurvis** [comment]
i think officially after considering it my opinion would be that the right thing to do is to remove the _deprecated_filters params and bump to 5.0.

**Comment by apurvis** [comment]
OK i pushed a commit that does away with `_deprecated_filters` that i'm officially proposing as 5.0.

There's no real reason to fear version bumps and sticking to the "if it breaks anything, you really have to bump the major version number" is pretty important i think for users (a lot of gems try to get around this rule in pursuit of some totally meaningless metric like keeping the version numbers low and it kind of sucks for the end users of those gems)

**Comment by jarthod** [comment]
Well, let's embrace the google chrome versioning scheme ;) I'm good with that

File: README.md [hunk 4]
```diff
@@ -6,6 +6,12 @@ Currently supports SurveyGizmo API **v4** (default) and **v3**.
 
 ## Versions
 
+### Major Changes in 5.x
+
+* BREAKING CHANGE: `.all` returns Enumerators, not arrays.  This may or may not break your code.
+* Feature: .all will automatically paginate responses for you with the :all_pages option (and it will also handle retries if you configure it)
+* Feature: Answer class to parse SurveyGizmo's sort of wild and wooly way of representing survey responses
```
Comment (jarthod) [inline README.md]: I think you should put the method name here

**Comment by apurvis** [comment]
maybe let it sit a bit til after christmas just in case i or you think of anything else you want to break while we have this opportunity.

**Comment by apurvis** [comment]
k i threw a Pester block around all the HTTParty calls; this explains in a nutshell why HTTParty kind of sucks compared to Faraday.

In a better world I would swap Httparty out but I'm not about to do that anytime soon.

**Comment by jarthod** [comment]
I'll merge this in master but wait a bit for the release, yes, so we can throw more breaking changes at it ;)

**Pull request merged**